File: modules/graphics/lib/aerox/graphics/shaders/GraphicsShader.cpp
#include "aerox/graphics/shaders/GraphicsShader.hpp"
#include <ranges>
#include <ashl/utils.hpp>
#include <iostream>
#include "aerox/graphics/descriptors/DescriptorLayoutBuilder.hpp"
#include "aerox/graphics/shaders/ShaderCompileError.hpp"
#include "ashl/tokenizer.hpp"
#include "ashl/parser.hpp"

namespace aerox::graphics
{
    std::unordered_map<std::string, Shared<GraphicsShader>> GraphicsShader::_shaders = {};

    GraphicsShader::GraphicsShader(ShaderManager* manager, const Shared<ashl::ModuleNode>& inVertex,
                                   const Shared<ashl::ModuleNode>& inFragment) : Shader(manager)
    {
        _vertexShader = inVertex;
        _fragmentShader = inFragment;
        _vertexShaderId = std::to_string(_vertexShader->ComputeHash());
        _fragmentShaderId = std::to_string(_fragmentShader->ComputeHash());
    }

    bool GraphicsShader::Bind(const vk::CommandBuffer& cmd, bool wait)
    {
        try
        {
            using namespace std::chrono_literals;
            if (wait)
            {
                _compiledShader.wait();
            }
            else if (_compiledShader.wait_for(0ms) != std::future_status::ready)
            {
                return false;
            }
            auto data = _compiledShader.get();
            std::vector<vk::ShaderEXT> shaders = {data.shaders[0].second,data.shaders[1].second};
            std::vector<vk::ShaderStageFlagBits> stages = {
                data.shaders[0].first, data.shaders[1].first
            };
        
            {
                auto stage = vk::ShaderStageFlagBits::eGeometry;
                cmd.bindShadersEXT(1, &stage, nullptr, GraphicsModule::dispatchLoader);
            }

            cmd.bindShadersEXT(stages, shaders, GraphicsModule::dispatchLoader);
            return true;
        }
        catch (ShaderCompileError& e)
        {
            std::cerr << "Shader Error:\n" << e.what() << std::endl;
            return false;
        }
    }

    std::string GraphicsShader::GetVertexShaderId() const
    {
        return _vertexShaderId;
    }

    std::string GraphicsShader::GetFragmentShaderId() const
    {
        return _fragmentShaderId;
    }

    std::shared_ptr<ashl::ModuleNode> GraphicsShader::GetVertexShader() const
    {
        return _vertexShader;
    }

    std::shared_ptr<ashl::ModuleNode> GraphicsShader::GetFragmentShader() const
    {
        return _fragmentShader;
    }

    Shared<GraphicsShader> GraphicsShader::FromFile(const std::filesystem::path& path)
    {
        using namespace ashl;

        const auto fullPath = std::filesystem::absolute(path).string();
        if (_shaders.contains(fullPath)) return _shaders[fullPath];
        auto manager = GraphicsModule::Get()->GetShaderManager();
        auto nodes = tokenize(fullPath);
        const auto ast = parse(nodes);
        auto vertex = extractScope(ast, ashl::EScopeType::Vertex);
        auto fragment = extractScope(ast, ashl::EScopeType::Fragment);
        resolveIncludes(vertex);
        resolveIncludes(fragment);
        resolveReferences(vertex);
        resolveReferences(fragment);
        auto shader = newShared<GraphicsShader>(manager, vertex, fragment);
        shader->Init();
        return shader;
    }

    void GraphicsShader::Init()
    {
        const auto thisShader = this->GetSharedDynamic<Shader>();


        ComputeResources({
            {vk::ShaderStageFlagBits::eVertex, _vertexShader}, {vk::ShaderStageFlagBits::eFragment, _fragmentShader}
        });
        
        _compiledShader = GetManager()->StartShaderCompilation(thisShader);
        
    }

    CompiledShader GraphicsShader::Compile(ShaderManager* manager)
    {
        auto vertexSpirv = manager->CompileAstToSpirv(_vertexShaderId, vk::ShaderStageFlagBits::eVertex, _vertexShader);
        auto fragmentSpirv = manager->CompileAstToSpirv(_fragmentShaderId, vk::ShaderStageFlagBits::eFragment,
                                                        _fragmentShader);


        DescriptorLayoutBuilder layoutBuilder{};
        std::vector<vk::PushConstantRange> pushConstantRanges{};
        std::map<uint32_t, DescriptorLayoutBuilder> builders{};
        CompiledShader shader{};

        for (auto& item : resources | std::views::values)
        {
            if (!builders.contains(item.set))
                builders.emplace(item.set,
                                 DescriptorLayoutBuilder{});

            builders.at(item.set).AddBinding(item.binding, item.type, item.stages, item.count, item.bindingFlags);
        }

        for (auto& item : pushConstants | std::views::values)
        {
            pushConstantRanges.emplace_back(item.stages, 0, static_cast<uint32_t>(item.size));
        }

        auto max = builders.empty() ? 0 : builders.rbegin()->first;
        std::vector<vk::DescriptorSetLayout> layouts{};
        layouts.reserve(max);
        for (auto i = 0; i < max + 1; i++)
        {
            auto newLayout = builders.contains(i) ? builders[i].Build() : DescriptorLayoutBuilder{}.Build();
            shader.descriptorLayouts.emplace(i, newLayout);
            layouts.push_back(newLayout);
        }


        auto device = manager->GetGraphicsModule()->GetDevice();
        {
            vk::PipelineLayoutCreateInfo createInfo{};
            createInfo.setSetLayouts(layouts);
            createInfo.setPushConstantRanges(pushConstantRanges);
        
            shader.pipelineLayout = device.createPipelineLayout(createInfo);
        }
        
        {
            auto createInfo = vk::ShaderCreateInfoEXT{
                {},
                vk::ShaderStageFlagBits::eVertex,
                vk::ShaderStageFlagBits::eFragment,
                vk::ShaderCodeTypeEXT::eSpirv,
                vertexSpirv.size() * sizeof(uint32_t),
                vertexSpirv.data(),
                "main",
                static_cast<unsigned int>(layouts.size()),
                layouts.data(),
                static_cast<unsigned int>(pushConstantRanges.size()),
                pushConstantRanges.data()
            };

            auto result = device.createShaderEXT(createInfo, nullptr, GraphicsModule::dispatchLoader);
            if (result.result != vk::Result::eSuccess)
            {
                throw std::runtime_error("Failed to compile shader");
            }

            shader.shaders.emplace_back(vk::ShaderStageFlagBits::eVertex, result.value);
        }

        {
            auto createInfo = vk::ShaderCreateInfoEXT{
                {},
                vk::ShaderStageFlagBits::eFragment,
                static_cast<vk::ShaderStageFlagBits>(0),
                vk::ShaderCodeTypeEXT::eSpirv,
                fragmentSpirv.size() * sizeof(uint32_t),
                fragmentSpirv.data(),
                "main",
                static_cast<unsigned int>(layouts.size()),
                layouts.data(),
                static_cast<unsigned int>(pushConstantRanges.size()),
                pushConstantRanges.data()
            };
            auto result = device.createShaderEXT(createInfo, nullptr, GraphicsModule::dispatchLoader);
            if (result.result != vk::Result::eSuccess)
            {
                throw std::runtime_error("Failed to compile shader");
            }

            shader.shaders.emplace_back(vk::ShaderStageFlagBits::eFragment, result.value);
        }


        manager->onDispose->Add([device,shader,layouts]
        {
           for (auto &descriptorSetLayout : layouts)
           {
               device.destroyDescriptorSetLayout(descriptorSetLayout);
           }

            device.destroyPipelineLayout(shader.pipelineLayout);
            
            for (const auto& shaderObject : shader.shaders | std::views::values)
            {
                device.destroyShaderEXT(shaderObject,nullptr,GraphicsModule::dispatchLoader);
            } 
        });
        return shader;
    }

    std::map<uint32_t, vk::DescriptorSetLayout> GraphicsShader::GetDescriptorSetLayouts()
    {
        return _compiledShader.get().descriptorLayouts;
    }

    vk::PipelineLayout GraphicsShader::GetPipelineLayout()
    {
        return _compiledShader.get().pipelineLayout;
    }
}

File: aerox.Runtime.Graphics/Shaders/GraphicsShader.cs
using aerox.Runtime.Graphics.Descriptors;
using rsl;
using rsl.Nodes;
using TerraFX.Interop.Vulkan;
using static TerraFX.Interop.Vulkan.Vulkan;
namespace aerox.Runtime.Graphics.Shaders;

public class GraphicsShader : Shader
{
    private string _vertexShaderId = "";
    private string _fragmentShaderId = "";
    private ModuleNode _vertexShader;
    private ModuleNode _fragmentShader;
    private Task<CompiledShader> _compiledShader;


    public override CompiledShader Compile(ShaderManager manager)
    {
        var vertexSpirv = manager.CompileAstToSpirv(_vertexShaderId, VkShaderStageFlags.VK_SHADER_STAGE_VERTEX_BIT,
            _vertexShader);
        var fragmentSpirv = manager.CompileAstToSpirv(_fragmentShaderId,
            VkShaderStageFlags.VK_SHADER_STAGE_FRAGMENT_BIT, _fragmentShader);

        var layoutBuilder = new DescriptorLayoutBuilder();
        List<VkPushConstantRange> pushConstantRanges = [];
        SortedDictionary<uint, DescriptorLayoutBuilder> builders = [];
        var shader = new CompiledShader();
        
        foreach (var (key, item) in Resources)
        {
            if (!builders.ContainsKey(item.Set))
            {
                builders.Add(item.Set,new DescriptorLayoutBuilder());
            }

            builders[item.Set].AddBinding(item.Binding, item.Type, item.Stages, item.Count, item.BindingFlags);
        }
        
        foreach (var (key, item) in PushConstants)
        {
            pushConstantRanges.Add(new VkPushConstantRange()
            {
                stageFlags = item.Stages,
                offset = 0,
                size = (uint)item.Size
            });
        }

        var max = builders.Count == 0 ? 0 : builders.Keys.Max();
        List<VkDescriptorSetLayout> layouts = [];

        for (uint i = 0; i < max + 1; i++)
        {
            var newLayout = builders.TryGetValue(i, out var value) ? value.Build() : new DescriptorLayoutBuilder().Build();
            shader.DescriptorLayouts.Add(i,newLayout);
            layouts.Add(newLayout);
        }

        var device = SGraphicsModule.Get().GetDevice();
        unsafe {
            fixed (VkDescriptorSetLayout * pSetLayouts = layouts.ToArray())
            {
                var setLayoutCount = (uint)layouts.Count;
                var pushConstantRangeCount = (uint)pushConstantRanges.Count;
                fixed (VkPushConstantRange * pPushConstants = pushConstantRanges.ToArray())
                {
                    {
                        var pipelineLayoutCreateInfo = new VkPipelineLayoutCreateInfo()
                        {
                            sType = VkStructureType.VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO,
                            setLayoutCount = setLayoutCount,
                            pSetLayouts = pSetLayouts,
                            pushConstantRangeCount = pushConstantRangeCount,
                            pPushConstantRanges = pPushConstants
                        };

                        var pipelineLayout = new VkPipelineLayout();
                        vkCreatePipelineLayout(device, &pipelineLayoutCreateInfo, null, &pipelineLayout);
                        shader.PipelineLayout = pipelineLayout;
                    }
                    fixed(byte * pName = "main"u8.ToArray())
                    {
                        {
                            var createInfo = new VkShaderCreateInfoEXT()
                            {
                                sType = VkStructureType.VK_STRUCTURE_TYPE_SHADER_CREATE_INFO_EXT,
                                stage = VkShaderStageFlags.VK_SHADER_STAGE_VERTEX_BIT,
                                nextStage = VkShaderStageFlags.VK_SHADER_STAGE_FRAGMENT_BIT,
                                codeType = VkShaderCodeTypeEXT.VK_SHADER_CODE_TYPE_SPIRV_EXT,
                                codeSize = (nuint)vertexSpirv.GetByteSize(),
                                pCode = vertexSpirv.GetData(),
                                pName = (sbyte*)pName,
                                pushConstantRangeCount = pushConstantRangeCount,
                                setLayoutCount = setLayoutCount,
                                pPushConstantRanges = pPushConstants,
                                pSetLayouts = pSetLayouts,
                                pSpecializationInfo = null,
                                pNext = null
                            };
                            var result = new VkShaderEXT();
                            if (VulkanExtensions.vkCreateShadersEXT(device, 1, &createInfo, null, &result) !=
                                VkResult.VK_SUCCESS)
                            {
                                throw new Exception("Failed to compile shader");
                            }
                        
                            shader.Shaders.Add(new Pair<VkShaderEXT,VkShaderStageFlags>(result,VkShaderStageFlags.VK_SHADER_STAGE_VERTEX_BIT));
                        }
                        
                        {
                            var createInfo = new VkShaderCreateInfoEXT()
                            {
                                sType = VkStructureType.VK_STRUCTURE_TYPE_SHADER_CREATE_INFO_EXT,
                                stage = VkShaderStageFlags.VK_SHADER_STAGE_FRAGMENT_BIT,
                                nextStage = 0,
                                codeType = VkShaderCodeTypeEXT.VK_SHADER_CODE_TYPE_SPIRV_EXT,
                                codeSize = (nuint)fragmentSpirv.GetByteSize(),
                                pCode = fragmentSpirv.GetData(),
                                pName = (sbyte*)pName,
                                pushConstantRangeCount = pushConstantRangeCount,
                                setLayoutCount = setLayoutCount,
                                pPushConstantRanges = pPushConstants,
                                pSetLayouts = pSetLayouts
                            };
                            var result = new VkShaderEXT();
                            if (VulkanExtensions.vkCreateShadersEXT(device, 1, &createInfo, null, &result) !=
                                VkResult.VK_SUCCESS)
                            {
                                throw new Exception("Failed to compile shader");
                            }
                        
                            shader.Shaders.Add(new Pair<VkShaderEXT,VkShaderStageFlags>(result,VkShaderStageFlags.VK_SHADER_STAGE_FRAGMENT_BIT));
                        }
                    }
                }
            }

            manager.BeforeDispose += () =>
            {
                vkDestroyPipelineLayout(device, shader.PipelineLayout, null);
                foreach (var (shaderObj,_ ) in shader.Shaders)
                {
                    VulkanExtensions.vkDestroyShaderEXT(device, shaderObj, null);
                }
            };

            return shader;
        }
    }

   


    public GraphicsShader(ModuleNode vertexShader,ModuleNode fragmentShader)
    {
        _vertexShader = vertexShader;
        _fragmentShader = fragmentShader;
        _vertexShaderId = vertexShader.GetHashCode().ToString();
        _fragmentShaderId = fragmentShader.GetHashCode().ToString();
    }

    public override void Init()
    {
        ComputeResources([new Pair<VkShaderStageFlags, ModuleNode>(VkShaderStageFlags.VK_SHADER_STAGE_VERTEX_BIT,_vertexShader),new Pair<VkShaderStageFlags, ModuleNode>(VkShaderStageFlags.VK_SHADER_STAGE_FRAGMENT_BIT,_fragmentShader)]);

        _compiledShader = SGraphicsModule.Get().GetShaderManager().StartShaderCompilation(this);
    }
    
    public override bool Bind(VkCommandBuffer cmd, bool wait = false)
    {
        if (wait && !_compiledShader.IsCompleted)
        {
            _compiledShader.Wait();
        }
        else if(!_compiledShader.IsCompleted)
        {
            return false;
        }

        cmd.UnBindShader(VkShaderStageFlags.VK_SHADER_STAGE_GEOMETRY_BIT);
        cmd.BindShaders(_compiledShader.Result.Shaders);

        return true;
    }

    public override Dictionary<uint, VkDescriptorSetLayout> GetDescriptorSetLayouts()
    {
        return _compiledShader.Result.DescriptorLayouts;
    }

    public override VkPipelineLayout GetPipelineLayout()
    {
        return _compiledShader.Result.PipelineLayout;
    }

    private static readonly Dictionary<string, GraphicsShader> GraphicsShaders = [];
    
    public static GraphicsShader FromFile(string filePath)
    {
        var fullPath = Path.GetFullPath(filePath);

        if (GraphicsShaders.ContainsKey(filePath)) return GraphicsShaders[fullPath];
        
        var tokenList = Tokenizer.Run(fullPath);

        var ast = Parser.Parse(ref tokenList);
        ast = ast.ResolveIncludes((node, moduleNode) =>
        {
            if (Path.IsPathRooted(node.File)) return Path.GetFullPath(node.File);

            return Path.GetFullPath(node.File,
                Directory.GetParent(node.SourceFile)?.FullName ?? Directory.GetCurrentDirectory());
        });
        ast.ResolveStructReferences();
        var vertexShader = ast.ExtractScope(ScopeType.Vertex);
        var fragmentShader = ast.ExtractScope(ScopeType.Fragment);

        var shader = new GraphicsShader(vertexShader, fragmentShader);
            
        GraphicsShaders.Add(fullPath,shader);
            
        shader.Init();
        return shader;
    }
}